File: cpp/solver.cpp
#include "solver.h"

#include <vector>
#include <iostream>
#include <cstring>

struct SolveGrid
{
    SolveGrid(const StarGrid& grid) :
        star_pos_len(grid.grid.size()),
        star_count(0),
        star_pos(new uint8_t[star_pos_len])
    {
        for (size_t i = 0; i < grid.grid.size(); ++i)
        {
            star_pos[i] = false;
        }
    }

    const size_t star_pos_len;
    size_t star_count;
    uint8_t* star_pos;

    ~SolveGrid()
    {
        if (star_pos != nullptr)
        {
            delete[] star_pos;
            star_pos = nullptr;
        }
    }
};

struct SolveGridState
{
    SolveGridState(const StarGrid& grid) :
        count_len(grid.dim_size),
        valid_len(grid.grid.size())
    {
        create_pointers();

        for (size_t i = 0; i < count_len; ++i)
        {
            star_shape_count[i] = 0;
            star_row_count[i] = 0;
            star_col_count[i] = 0;
            free_shape_count[i] = static_cast<uint8_t>(grid.shape_indices[i].size());
            free_row_count[i] = grid.dim_size;
            free_col_count[i] = grid.dim_size;
        }

        for (size_t i = 0; i < valid_len; ++i)
        {
            valid_cells[i] = true;
        }
    }

    SolveGridState(const SolveGridState& other) :
        count_len(other.count_len),
        valid_len(other.valid_len)
    {
        create_pointers();
        update_vals(other);
    }

    SolveGridState& operator=(const SolveGridState& other)
    {
        if (this == &other)
        {
            return *this;
        }

        if (count_len != other.count_len || valid_len != other.valid_len)
        {
            delete_pointers();
            create_pointers();
        }

        update_vals(other);
        return *this;
    }

    const size_t count_len;
    const size_t valid_len;

    uint8_t* star_shape_count;
    uint8_t* star_row_count;
    uint8_t* star_col_count;
    uint8_t* free_shape_count;
    uint8_t* free_row_count;
    uint8_t* free_col_count;
    uint8_t* valid_cells;

    void update_vals(const SolveGridState& other)
    {
        memcpy(star_shape_count, other.star_shape_count, sizeof(uint8_t) * count_len);
        memcpy(star_row_count, other.star_row_count, sizeof(uint8_t) * count_len);
        memcpy(star_col_count, other.star_col_count, sizeof(uint8_t) * count_len);
        memcpy(free_shape_count, other.free_shape_count, sizeof(uint8_t) * count_len);
        memcpy(free_row_count, other.free_row_count, sizeof(uint8_t) * count_len);
        memcpy(free_col_count, other.free_col_count, sizeof(uint8_t) * count_len);
        memcpy(valid_cells, other.valid_cells, sizeof(uint8_t) * valid_len);
    }

    ~SolveGridState()
    {
        delete_pointers();
    }

private:
    void create_pointers()
    {
        star_shape_count = new uint8_t[count_len];
        star_row_count = new uint8_t[count_len];
        star_col_count = new uint8_t[count_len];
        free_shape_count = new uint8_t[count_len];
        free_row_count = new uint8_t[count_len];
        free_col_count = new uint8_t[count_len];
        valid_cells = new uint8_t[valid_len];
    }

    void delete_pointers()
    {
        if (star_shape_count != nullptr) delete[] star_shape_count;
        if (star_row_count != nullptr) delete[] star_row_count;
        if (star_col_count != nullptr) delete[] star_col_count;
        if (free_shape_count != nullptr) delete[] free_shape_count;
        if (free_row_count != nullptr) delete[] free_row_count;
        if (free_col_count != nullptr) delete[] free_col_count;
        if (valid_cells != nullptr) delete[] valid_cells;
    }
};

bool add_star(const StarGrid& grid, SolveGrid& base, std::vector<SolveGridState>& base_states, size_t ind, SolvedState& state_out)
{
    // Create a copy of the state
    SolveGridState& state = base_states[base.star_count + 1];
    state.update_vals(base_states[base.star_count]);

    // Extract the current shape ID
    const size_t sid = grid.grid[ind];

    // Convert ind to row/col
    const size_t row = grid.ind_to_row(ind);
    const size_t col = grid.ind_to_col(ind);

    // Mark the grid star as added
    state.star_shape_count[sid] += 1;
    state.star_row_count[row] += 1;
    state.star_col_count[col] += 1;

    // Loop over boundary cells around the added cell value
    {
        const size_t row_min = (row > 0) ? row - 1 : 0;
        const size_t row_max = std::min(row + 2, grid.dim_size);

        const size_t col_min = (col > 0) ? col - 1 : 0;
        const size_t col_max = std::min(col + 2, grid.dim_size);

        for (size_t row_i = row_min; row_i < row_max; ++row_i)
        {
            for (size_t col_i = col_min; col_i < col_max; ++col_i)
            {
                const size_t ind_rc = grid.rc_to_ind(row_i, col_i);
                if (state.valid_cells[ind_rc])
                {
                    state.valid_cells[ind_rc] = false;
                    state.free_shape_count[grid.grid[ind_rc]] -= 1;
                    state.free_row_count[row_i] -= 1;
                    state.free_col_count[col_i] -= 1;
                }
            }
        }
    }

    // Check if the row is full
    if (state.star_row_count[row] == grid.stars_per_object)
    {
        for (size_t k = 0; k < grid.dim_size; ++k)
        {
            const size_t indi = grid.rc_to_ind(row, k);
            if (state.valid_cells[indi])
            {
                state.valid_cells[indi] = false;
                state.free_shape_count[grid.grid[indi]] -= 1;
                state.free_row_count[row] -= 1;
                state.free_col_count[k] -= 1;
            }
        }
    }

    // Check if the column is full
    if (state.star_col_count[col] == grid.stars_per_object)
    {
        for (size_t k = 0; k < grid.dim_size; ++k)
        {
            const size_t indi = grid.rc_to_ind(k, col);
            if (state.valid_cells[indi])
            {
                state.valid_cells[indi] = false;
                state.free_shape_count[grid.grid[indi]] -= 1;
                state.free_row_count[k] -= 1;
                state.free_col_count[col] -= 1;
            }
        }
    }

    // Check if the shape is now full
    if (state.star_shape_count[sid] == grid.stars_per_object)
    {
        const std::vector<size_t>& shape_is = grid.shape_indices[sid];
        const size_t gs = shape_is.size();
        for (size_t i = 0; i < gs; ++i)
        {
            const size_t indi = shape_is[i];
            if (state.valid_cells[indi])
            {
                state.valid_cells[indi] = false;
                state.free_shape_count[sid] -= 1;
                state.free_row_count[grid.ind_to_row(indi)] -= 1;
                state.free_col_count[grid.ind_to_col(indi)] -= 1;
            }
        }
    }

    // Check for invalid grid solution to trim constraints
    for (size_t i = 0; i < grid.dim_size; ++i)
    {
        if (state.free_col_count[i] + state.star_col_count[i] < grid.stars_per_object)
        {
            return false;
        }
        else if (state.free_row_count[i] + state.star_row_count[i] < grid.stars_per_object)
        {
            return false;
        }
        else if (state.free_shape_count[i] + state.star_shape_count[i] < grid.stars_per_object)
        {
            return false;
        }
    }

    // Increment and add the star count value
    base.star_count += 1;
    base.star_pos[ind] = true;

    // Check for a solution
    if (base.star_count == grid.target_stars)
    {
        state_out.star_pos = base.star_pos;
        base.star_pos = nullptr;
        state_out.valid_pos = state.valid_cells;
        state.valid_cells = nullptr;
        return true;
    }

    // Loop through future indices to add potential cells

    for (size_t i = ind; i < grid.grid.size(); ++i)
    {
        if (state.valid_cells[i])
        {
            const size_t ri = grid.ind_to_row(i);
            if (ri > 0 && state.star_row_count[ri - 1] < grid.stars_per_object)
            {
                break;
            }

            if (add_star(grid, base, base_states, i, state_out))
            {
                return true;
            }
        }
    }

    base.star_count -= 1;
    base.star_pos[ind] = false;

    if (state.star_row_count[row] == 0)
    {
        SolveGridState& prev_state = base_states[base.star_count];
        prev_state.valid_cells[ind] = false;
        prev_state.free_row_count[row] -= 1;
        prev_state.free_col_count[col] -= 1;
        prev_state.free_shape_count[sid] -= 1;
    }

    return false;
}

bool solve_battle_grid(const StarGrid& grid, SolvedState& state_out)
{
    // Define the starting base grid and states
    SolveGrid base(grid);
    std::vector<SolveGridState> base_states(grid.dim_size * grid.stars_per_object + 1, SolveGridState(grid));

    // Define the basic loop to check the first row
    for (size_t i = 0; i < grid.dim_size; ++i)
    {
        // Add the star and check for solution
        if (add_star(grid, base, base_states, i, state_out))
        {
            return true;
        }
    }

    // Return false on no solution
    return false;
}

File: rust/solver.rs
use crate::stargrid::StarGrid;
use crate::stargrid::SolvedState;

//#[derive(Clone)]
#[derive(Debug)]
struct SolveGrid
{
    star_count: u8,
    star_pos: Vec<bool>,
}

#[derive(Clone, Debug)]
struct SolveGridState
{
    star_shape_count: Vec<u8>,
    star_row_count: Vec<u8>,
    star_col_count: Vec<u8>,
    free_shape_count: Vec<u8>,
    free_row_count: Vec<u8>,
    free_col_count: Vec<u8>,
    valid_cells: Vec<bool>
}

fn create_main_solve(grid: &StarGrid) -> SolveGrid
{
    return SolveGrid
    {
        star_count: 0,
        star_pos: (0..grid.grid.len()).map(|_| false).collect::<Vec<bool>>()
    };
}

fn create_solve_state(grid: &StarGrid) -> SolveGridState
{
    let star_count: Vec<u8> = (0..grid.dim_size).map(|_| 0).collect::<Vec<u8>>();
    let free_count: Vec<u8> = (0..grid.dim_size).map(|_| grid.dim_size as u8).collect::<Vec<u8>>();

    let free_shape_count: Vec<u8> = grid.shape_indices.iter().map(|v| v.len() as u8).collect::<Vec<u8>>();
    let valid_cells: Vec<bool> = (0..grid.grid.len()).map(|_| true).collect::<Vec<bool>>();

    return SolveGridState
    {
        valid_cells: valid_cells,
        star_shape_count: star_count.clone(),
        star_row_count: star_count.clone(),
        star_col_count: star_count.clone(),
        free_shape_count: free_shape_count,
        free_row_count: free_count.clone(),
        free_col_count: free_count.clone()
    };
}

fn row_col_iter(grid: &StarGrid, u: usize) -> std::ops::Range<usize>
{
    return (if u > 0 { u - 1 } else { u })..std::cmp::min(grid.dim_size, u + 2);
}

fn add_star(grid: &StarGrid, solver: &mut SolveGrid, state_prev: &mut SolveGridState, ind: usize) -> std::option::Option<SolvedState>
{
    // Create a copy of the state
    let mut state: SolveGridState = (*state_prev).clone();

    // Check if there is already a star in the given position
    if solver.star_pos[ind] || !state.valid_cells[ind]
    {
        println!("This shouldn't ever happen!");
        return None;
    }

    // Extract the added star's grid shape ID
    let sid: u8 = grid.grid[ind];
    let sid_ind: usize = sid as usize;

    // Convert ind to row/col
    let row: usize = grid.ind_to_row(ind);
    let col: usize = grid.ind_to_col(ind);

    // Mark the grid star as added
    state.star_shape_count[sid as usize] += 1;
    state.star_row_count[row] += 1;
    state.star_col_count[col] += 1;

    // Loop over boundary cells around the added cell value
    for row_i in row_col_iter(grid, row)
    {
        for col_i in row_col_iter(grid, col)
        {
            let ind_rc = grid.rowcol_to_ind(row_i, col_i);
            if state.valid_cells[ind_rc]
            {
                state.valid_cells[ind_rc] = false;
                state.free_shape_count[grid.grid[ind_rc] as usize] -= 1;
                state.free_row_count[row_i] -= 1;
                state.free_col_count[col_i] -= 1;
            }
        }
    }

    // Check if the shape is now full
    if state.star_shape_count[sid_ind] == grid.target_star_count * grid.dim_size as u8
    {
        for i in &grid.shape_indices[sid_ind]
        {
            let ii: usize = *i;
            if state.valid_cells[ii]
            {
                state.valid_cells[ii] = false;
                state.free_shape_count[sid_ind] -= 1;
                state.free_row_count[grid.ind_to_row(ii)] -=1;
                state.free_col_count[grid.ind_to_col(ii)] -=1;
            }
        }
    }

    // Check if the row is full
    if state.star_row_count[row] == grid.target_star_count
    {
        for k in 0..grid.dim_size
        {
            let ii: usize = grid.rowcol_to_ind(row, k);

            if state.valid_cells[ii]
            {
                state.valid_cells[ii] = false;
                state.free_shape_count[grid.grid[ii] as usize] -= 1;
                state.free_row_count[row] -= 1;
                state.free_col_count[k] -= 1;
            }
        }
    }

    // Check if the column is full
    if state.star_col_count[col] == grid.target_star_count
    {
        for k in 0..grid.dim_size
        {
            let ii: usize = grid.rowcol_to_ind(k, col);

            if state.valid_cells[ii]
            {
                state.valid_cells[ii] = false;
                state.free_shape_count[grid.grid[ii] as usize] -= 1;
                state.free_row_count[k] -= 1;
                state.free_col_count[col] -= 1;
            }
        }
    }

    // Check for invalid grid solution to trim constraints
    for i in 0..grid.dim_size
    {
        if state.free_col_count[i] + state.star_col_count[i] < grid.target_star_count
        {
            return None;
        }
        else if state.free_row_count[i] + state.star_row_count[i] < grid.target_star_count
        {
            return None;
        }
        else if state.free_shape_count[i] + state.star_shape_count[i] < grid.target_star_count
        {
            return None;
        }
    }

    // Increment and add the star count value
    solver.star_count += 1;
    solver.star_pos[ind] = true;

    // Check for a solution
    if solver.star_count == grid.target_star_count * grid.dim_size as u8
    {
        return Some(SolvedState
        {
            star_pos: solver.star_pos.clone(),
            valid_cell: state.valid_cells.clone()
        });
    }

    // Loop through future indices to add potential cells
    for i in ind..grid.grid.len()
    {
        if state.valid_cells[i]
        {
            let ri = grid.ind_to_row(i);
            if ri > 0 && state.star_row_count[ri - 1] < grid.target_star_count
            {
                break;
            }

            let rval = add_star(grid, solver, &mut state, i);
            if rval.is_some()
            {
                return rval;
            }
        }
    }

    // Remove the star and uncheck the star position from the state grid
    solver.star_count -= 1;
    solver.star_pos[ind] = false;

    // Invalidate the current cell from the previous state if no possible solutions were found
    if state.star_row_count[row] == 0
    {
        state_prev.valid_cells[ind] = false;
        state_prev.free_row_count[row] -= 1;
        state_prev.free_col_count[col] -= 1;
        state_prev.free_shape_count[sid_ind] -= 1;
    }

    // Return no solution on failure
    return None;
}

pub fn solve_battle_grid(grid: &StarGrid) -> std::option::Option<SolvedState>
{
    // Create the base object and solve state
    let mut base = create_main_solve(grid);
    let mut base_state = create_solve_state(grid);

    // Iterate over the first row of values
    for i in 0..grid.dim_size
    {
        // Check the return value for the solution
        let ret = add_star(grid, &mut base, &mut base_state, i);

        // Detect if the return value is successful; if so, return
        if ret.is_some()
        {
            return ret;
        }
    }

    // Return no solution on failure
    return None;
}
